kele.el

```diff
@@ -675,42 +755,35 @@ after a certain amount of time.
 
 If PORT is nil, a random port will be chosen.
 
-Returns the proxy process."
+Returns an alist with keys `proc', `timer', and `port'.
+
+If CONTEXT already has a proxy process active, this function returns the
+existing process *regardless of the value of PORT*."
   (interactive (list (completing-read "Start proxy for context: " #'kele--contexts-complete)
                      :port nil
                      :ephemeral t))
-  ;; TODO: Throw error if proxy already active for context
-  (kele--with-progress (format "Starting proxy server process for `%s'..." context)
-    (let* ((selected-port (or port (kele--random-port)))
-           (key (intern context))
-           (proc (kele--proxy-process context :port selected-port))
-           (cleanup (when ephemeral
-                      (run-with-timer kele-proxy-ttl nil #'kele-proxy-stop context)))
-           ;; TODO: Define a struct for this
-           (entry `((proc . ,proc)
-                    (timer . ,cleanup)
-                    (port . ,selected-port))))
-      (add-to-list 'kele--context-proxy-ledger `(,key . ,entry))
-      entry)))
-
-(defun kele--proxy-enabled-p (context)
-  "Return non-nil if proxy server process active for CONTEXT."
-  (alist-get (intern context) kele--context-proxy-ledger))
+  (proxy-start kele--global-proxy-manager context :port port :ephemeral ephemeral)
+  (list (cons 'proc (proxy-get kele--global-proxy-manager context :wait nil))
+        (cons 'timer (cdr (assoc context (oref kele--global-proxy-manager timers))))
+        (cons 'port port)))
 
 (defun kele-proxy-toggle (context)
   "Start or stop proxy server process for CONTEXT."
   (interactive (list (completing-read
                       "Start/stop proxy for context: "
                       #'kele--contexts-complete)))
   (funcall
-   (if (kele--proxy-enabled-p context) #'kele-proxy-stop #'kele-proxy-start)
+   (if (proxy-active-p kele--global-proxy-manager context)
+       #'kele-proxy-stop
+     #'kele-proxy-start)
    context))
 
 (cl-defun kele--ensure-proxy (context)
-  "Return a proxy process for CONTEXT, creating one if needed."
-  (if-let* ((entry (alist-get (intern context) kele--context-proxy-ledger)))
-      entry
-    (kele-proxy-start context)))
+  "Return a proxy process for CONTEXT, creating one if needed.
+
+Returns the port of the proxy process."
+  (kele-proxy-start context)
+  (cdr (assoc context (oref kele--global-proxy-manager ports))))
```

Comment:
`--ensure-proxy` is largely vestigial at this point. We should get rid of it.

To do so, we'll need `kele--proxy-manager::proxy-get` to return not only the process, but also the port (and throw in the timer too, why not).

Basically, have it return the historical "proxy ledger entry" alist.